### PR TITLE
Development

### PR DESCRIPTION
**Pull request by DanielGreibe**

### CODE REVIEW

File: src/main/java/MongoConnection.java [hunk 1]
```diff
@@ -3,27 +3,51 @@
 import com.mongodb.client.MongoClients;
 import com.mongodb.client.MongoDatabase;
 
+/**
+ * jdoc description
+ * */
 public class MongoConnection
 {
-    private static MongoConnection mongoConnection = null;
+    private static boolean TestEnvironment = true;
```
Comment (cbudtz) [inline src/main/java/MongoConnection.java]: Use Environment Variable - eks System.getEnv("NITRO_TESTENVIRONMENT);

File: src/main/java/MongoConnection.java [hunk 2]
```diff
@@ -3,27 +3,51 @@
...
+    private static MongoConnection mongoTestConnection = null;
```
Comment (cbudtz) [inline src/main/java/MongoConnection.java]: Code will be cleaner if there is only one mongoConnection (Which is a testConnection when environment says so)

File: src/main/java/MongoConnection.java [hunk 3]
```diff
@@ -3,27 +3,51 @@
...
+    private static MongoConnection mongoProductionConnection = null;
 
     public static MongoConnection getInstance(String database)
     {
-        if (mongoConnection == null)
+        if(TestEnvironment)
         {
-            mongoConnection = new MongoConnection(database);
+            if (mongoTestConnection == null)
```
Comment (cbudtz) [inline src/main/java/MongoConnection.java]: Look at comment for line 12

File: src/main/java/MongoDBDAO.java [hunk 4]
```diff
@@ -1,139 +1,140 @@
...
 
 public class MongoDBDAO implements WeatherstationDAO {
 
+
+
     @Override
-    public String InsertToDatabase(String jsonMessage) {
-        MongoCollection<Document> collection = MongoConnection.getInstance("AgricircleDB").getDatabase().getCollection("Weatherstation1");
+    public String setWeatherstationData(String jsonMessage, String weatherstationName) throws ServerException {
+        //For Production:
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Again - let environment variable control this

File: src/main/java/MongoDBDAO.java [hunk 5]
```diff
@@ -1,139 +1,140 @@
...
+        MongoCollection<Document> collection = MongoConnection.getInstance("AgricircleDB").getDatabase().getCollection("Weatherstation1");
         try {
             collection.insertOne(Document.parse(jsonMessage));
             return "Database updated";
-        } catch (MongoException mwe) {
-            //  Block of code to handle error
-            return mwe.getMessage();
-            //throw new DBException("Soe")
+        } catch (MongoWriteException | MongoWriteConcernException e) {
+            throw new ServerException(e.getMessage());
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Subclass your own exception - Don't use rmi.ServerException

File: src/main/java/MongoDBDAO.java [hunk 6]
```diff
@@ -1,139 +1,140 @@
...
+                "payload_fields.outside_humidity",
+                "payload_fields.barometer_data",
+                "payload_fields.rain_rate"),
+                excludeId()));
+        if(contentType.equals("netcdf"))
+        {
+            //TODO Convert JSON to netcdf and return it.
+            return "Not Yet Implemented. Use contenttype = json";
+        }
+        if(iterable != null) {
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Use ObjectMapper - Don't reimplement this

File: src/main/java/MongoDBDAO.java [hunk 7]
```diff
@@ -1,139 +1,140 @@
...
-            return returnString.toString();
-        } catch (MongoException mwe) {
-            //  Block of code to handle errors
-            return mwe.getMessage();
+            returnString.delete(returnString.length()-1 , returnString.length());
+            returnString.append("]");
         }
+        if(returnString.length() == 0)
+        {
+            returnString.append("We couldn't find any data with the given criteria");
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Dont hardcode strings

File: src/main/java/MongoDBDAO.java [hunk 8]
```diff
@@ -1,139 +1,140 @@
...
+        }
+        return returnString.toString();
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Redundant toString on String

File: src/main/java/MongoDBDAO.java [hunk 9]
```diff
@@ -1,139 +1,140 @@
...
-            } else {
-                return "Could not find any data with the given criterias";
-            }
-        } catch (MongoException mwe) {
-            //  Block of code to handle errors
-            return mwe.getMessage();
-        }
-    }
+    public String getWeatherstationData(String stringDate , String weatherstationName , String contentType){
+        //For Production:
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: DRY  - dont repeat yourself

File: src/main/java/MongoDBDAO.java [hunk 10]
```diff
@@ -1,139 +1,140 @@
...
-            return mwe.getMessage();
-            //throw new DBException("Soe")
+        } catch (MongoWriteException | MongoWriteConcernException e) {
+            throw new ServerException(e.getMessage());
         }
     }
 
     @Override
-    public String findAllFromDatabase() {
+    public String getWeatherstationData(String weatherstationName , String contentType)
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Return Typed Object

File: src/main/java/MongoDBDAO.java [hunk 11]
```diff
@@ -1,139 +1,140 @@
...
+                        "payload_fields.solar_radiation",
+                        "payload_fields.outside_temperature",
+                        "payload_fields.outside_humidity",
+                        "payload_fields.barometer_data",
+                        "payload_fields.rain_rate"),
+                        excludeId()));
+
+        if(contentType!= null && contentType.equals("netcdf"))
         {
-            return "We couldn't find any data with the given criteria";
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Throw NotImplementedException instead

File: src/main/java/MongoDBDAO.java [hunk 12]
```diff
@@ -1,139 +1,140 @@
...
+                returnString.delete(returnString.length() - 1, returnString.length());
+                returnString.append("]}");
+            }
         }
         return returnString.toString();
     }
 
-    private static String DateToHexString(int year, int month, int day)
-    {
+    private static String DateToHexString(int year, int month, int day) throws WrongDateFormatException {
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: What is this for? Is it neccessary? Looks wrong.

File: src/main/java/MongoDBDAO.java [hunk 13]
```diff
@@ -1,139 +1,140 @@
...
-            throw new InputMismatchException("Date should be earlier than current date");
+            throw new WrongDateFormatException("Date should be earlier than current date");
         }
         long time = (calendar.getTimeInMillis() / 1000);
         return Long.toHexString(time) + "0000000000000000";
     }
 
-    private ObjectId DateToObjectId(String stringDate)
-    {
+    private static ObjectId DateToObjectId(String stringDate) throws WrongDateFormatException {
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Why are you making ObjectIds from dates?

File: src/main/java/WeatherstationDAO.java [hunk 14]
```diff
@@ -1,12 +1,10 @@
+import java.rmi.ServerException;
```
Comment (cbudtz) [inline src/main/java/WeatherstationDAO.java]: Make your own exception

File: src/main/java/WeatherstationEndpoint.java [hunk 15]
```diff
@@ -1,52 +1,71 @@
...
+    public String setWeatherstationData(String message, @PathParam("weatherstation") String weatherstationName)
     {
+        System.out.println("Kald setWeatherstationData(message , weatherstationName) med " + message + " og " + weatherstationName);
+        try
+        {
+            return database.setWeatherstationData(message , weatherstationName);
+        }
+        catch(ServerException se)
+        {
+            return se.getMessage();
```
Comment (cbudtz) [inline src/main/java/WeatherstationEndpoint.java]: Do NOT convert an exception to a string - this returns 200 OK - which is wrong

Comment (cbudtz) [reply src/main/java/WeatherstationEndpoint.java]: ExceptionMapper is the rigth solution. Should return a meaningful errorCodehttps://howtodoinjava.com/jersey/jaxrs-jersey-exceptionmapper/

File: src/main/java/WeatherstationEndpoint.java [hunk 16]
```diff
@@ -1,52 +1,71 @@
...
     /**
-     * Provides all data stored in the database.
+     * Provides all data stored in the database from the weather station stated in the weatherstationName parameter.
+     * @param weatherstationName the name of the weather station to get data from.
+     * @param contentType the type of content the API produces. Can be netcdf or json. Default is json.
      * @return a JSON string with all data in the database.
      */
     @GET
-    public String getWeatherstationData()
+    public String getWeatherstationData(@PathParam("weatherstation") String weatherstationName , @QueryParam("contenttype") String contentType)
```
Comment (cbudtz) [inline src/main/java/WeatherstationEndpoint.java]: Use Enum for contentType - https://www.w3schools.com/java/java_enums.asp

File: src/main/java/MongoDBDAO.java [hunk 17]
```diff
@@ -1,139 +1,140 @@
-import com.mongodb.ConnectionString;
```
Comment (cbudtz) [inline src/main/java/MongoDBDAO.java]: Have a look at https://github.com/diplomit-dtu/diplomBackend/blob/master/src/main/java/data/mongoImpl/MongoCourseDAO.java for a simpler approach. Some things are overcomplicated

File: src/main/java/MongoDBDAO.java (hunk 5 above)
Comment (DanielGreibe) [reply src/main/java/MongoDBDAO.java]: I'm not sure what you mean by 'Subclass your own exception'
Could you elaborate?

File: src/main/java/MongoDBDAO.java (hunk 8 above)
Comment (DanielGreibe) [reply src/main/java/MongoDBDAO.java]: returnString is a StringBuilder object, which is the reason i'm using toString on it.

File: src/main/java/MongoDBDAO.java (hunk 13 above)
Comment (DanielGreibe) [reply src/main/java/MongoDBDAO.java]: Tomasz wanted the endpoint to return data between a given date, and the current date.
One way to filter is taking advantage of the generated _id in Mongo.
The _id consists of a field containing a timestamp. Therefore i take the date given as part of the path, convert it to a ObjectID and use it as a filter.

File: src/main/java/MongoDBDAO.java (hunk 12 above)
Comment (DanielGreibe) [reply src/main/java/MongoDBDAO.java]: "this" being having the method throw WrongDateFormatException?

**Pull request merged**